### DISCUSSION TITLE
Sensor used for forecasting in the Energy dashboard

### DISCUSSION THREAD

**Discussion by nicopret1** [BJReplay/ha-solcast-solar#167] [Q&A] [answered]
I want to create custom forecasting data that I can use for forecasting my solar in the energy dashboard, but it is not clear to me which of the sensor/s are used to display the forecasting trend for the day.

Also what will be required for Home Assistant to recognise my custom forecasting sensor? Any help or ideas will be appreciated.

<img width="499" alt="Screenshot 2024-09-10 at 09 19 28" src="https://github.com/user-attachments/assets/8db9727b-21b0-4941-904e-84c940b78e97">

**Comment by gcoan** [comment]
You don't need to create a custom sensor for the energy dashboard, you simply need to configure the energy dashboard to include your solcast solar forecast - see https://github.com/BJReplay/ha-solcast-solar?tab=readme-ov-file#set-up-ha-energy-dashboard-settings

**Comment by autoSteve** [comment]
I assume you want to take Solcast sensor data, manipulate it, then feed it to the energy dashboard.

~~I'm pretty certain this is not possible.~~ This is not possible. I double checked the energy integration code.

The way this integration sends the forecast there is from `energy.py`, which sends an energy dashboard compatible Python dictionary that has previously been built when forecasts are read from Solcast. It does not come from a sensor.

A snip of the business end of the code:

```
async def async_get_solar_forecast(hass: HomeAssistant, config_entry_id: str):
    """Get solar forecast for a config entry ID."""

    coordinator: SolcastUpdateCoordinator = hass.data[DOMAIN][config_entry_id]

    return coordinator.get_energy_tab_data()
```

What do you mean by  "custom forecasting data"? What is it that you are trying to achieve?

**Comment by autoSteve** [comment, marked as answer]
> Also what will be required for Home Assistant to recognise my custom forecasting sensor?

With respect (and a little humour),

* Ditch the sensor idea.
* Fork BJReplay/ha-solcast-solar.
* Learn Python if you don't know it.
* Alter the `Solcast._data_energy` dict data after it has been built to suit the need (or alter the `_data_forecasts` data before the `_data_energy` gets created like we do with dampening). Please conform with HASS coding style guidelines.
* Then add a release to your fork (using the same code version number forked). The download ZIP will get automatically created.
* Then add a custom HACS repo for your fork and download over the top of this integration.
* Create a PR if you think others could benefit, likely adding an option to the config dialogue to enable your customisations.

Easy. 😂

What could possibly go wrong in future? 🤕

Or I could have a go at it if it's an idea that others would likely benefit from.

**Reply by nicopret1** [reply]
Thank you for the quick response and suggestions @autoSteve. It will be a steep learning curve for me, but I can always try since you say it is easy 😂.

>>I assume you want to take Solcast sensor data, manipulate it, then feed it to the energy dashboard.
Yes, correct, my idea is to factor in aspects such as shading during certain times of the day.

**Reply by autoSteve** [reply]
I have gone down this path. (I deleted my branch in disgust. Don't go there... 🤣)

To satisfy my hopes and dreams I hope you pull this off. I failed because on my utility there is a grid export limit of 5kW. So is this limiting due to the inverters performing solar curtailing, or the impact of shade? To find all this out, I tried to get sensors outside of the Solcast integration, which did not go well.

An add-in can get anything it likes, because it is running with supervisor context. An integration can not, so it needs to resort to kludgy internal API calls, which need a user-supplied API token, which is when the branch got deleted with a sigh. If it's too complex for me to manage, then it definitely is for your average integration user.

So I'm back to HA automations that can tell the Solcast integration to adjust dampening factors based on expected shading at different times of year. It ain't perfect, but it's good enough. And it adjusts the energy dashboard forecast for you.

@Nilogax does this (https://github.com/BJReplay/ha-solcast-solar/discussions/120), and was threatening to do a write-up of his set-up.

**Reply by autoSteve** [reply]
(BTW, this integration will never become an add-in. There are too many deployment variants in use today, like unsupervised Docker. Another alternative I considered was an adjunctive add-in for this integration. That idea took off like a wrought iron hang glider at its early stages... but it's another alternative for you to consider.)

**Reply by Nilogax** [reply]
Just need another rainy day (or two) and I'll have something to share.

**Reply by autoSteve** [reply]
Aren't you in the UK, @Nilogax? Getting a rainy day should be like shooting fish in a barrel, even in summer... 😂

**Reply by Nilogax** [reply]
I'm talking a proper rainy day, not just a spot of drizzle. Anyhow here's the long range forecast... 

![Screenshot_20240910-141646](https://github.com/user-attachments/assets/c2f9a1c8-aa19-4cef-baaa-936e7bdac608)

**Reply by autoSteve** [reply]
Damn. It'll be a while... Just the odd surprised sun. 🫨

**Comment by zhouska** [comment]
I use a custom sensor for this exact purpose. It calculates dampening values using ``weather.forecast_home`` entity for conditions like cloudy, fog, hail, pouring and snowy. I have assigned some manual and arbitrary 'weights' (trial and error) to these conditions to recalculate the dampening values based on the output from the entity.

Then I run the solcast forecast twice a day now, at midnight to have a look at the day ahead, and then later in the evening, to have a look at the next day (and kick off necessary automations).

Not the best of ideas, but it helps me a lot, as I'd rather be conservative in my approach and charge my batteries from grid in off-peak hours in preparation for these 'dark' days where panels are covered with snow, for instance. My location/altitude dictates it.

**Reply by BJReplay** [reply]
> It calculates dampening values using `weather.forecast_home` entity for conditions like cloudy, fog, hail, pouring and snowy. I have assigned some manual and arbitrary 'weights' (trial and error) to these conditions to recalculate the dampening values based on the output from the entity.

This is fantastic.

If you could, sharing your code could be a great starting point for others to adapt from their forecast as well.

**Reply by zhouska** [reply]
> > It calculates dampening values using `weather.forecast_home` entity for conditions like cloudy, fog, hail, pouring and snowy. I have assigned some manual and arbitrary 'weights' (trial and error) to these conditions to recalculate the dampening values based on the output from the entity.
> 
> This is fantastic.
> 
> If you could, sharing your code could be a great starting point for others to adapt from their forecast as well.

Yes, no biggie, here it is:

```
- trigger:
    - platform: time
      at: "00:00:00"
    - platform: time
      at: "17:45:00"
  action:
    service: weather.get_forecasts
    target:
      entity_id: weather.forecast_home
    data:
      type: hourly
    response_variable: my_forecast
  sensor:
    - name: Daily dampening values from hourly forecast conditions
      unique_id: daily_dampening_values_from_hourly_forecast_conditions
      state: "{{ now().isoformat() }}"
      attributes:
        dampening: >
          {% set ns= namespace() %}
          {% set ns.conditions = '' %}
          {% set ns.partlycloudy = 0.70 %}
          {% set ns.cloudy = 0.50 %}
          {% set ns.fog = 0.35 %}
          {% set ns.hail = 0.45 %}
          {% set ns.rainy = 0.45 %}
          {% set ns.pouring = 0.35 %}
          {% set ns.windy = 0.65 %}
          {% set ns.snowy = 0.25 %}
          {% set ns.sunny = 1 %}
          {% set ns.exceptional = 1 %}
          {% set ns.multiplicator = "" %}
          {% set ns.separator = "," %}
          {%- for hour in range(0, 24) %}
            {%- if loop.last -%}
                {%- if my_forecast['weather.forecast_home'].forecast[hour].condition == 'partlycloudy' %}
                  {% set ns.multiplicator = ns.partlycloudy/my_forecast['weather.forecast_home'].forecast[hour].cloud_coverage*100 %}
                  {% if ns.multiplicator <= 1 %}
                  {% set ns.conditions = ns.conditions + ns.partlycloudy*ns.multiplicator|round(2)|string  %}
                  {% else %}
                  {% set ns.conditions = ns.conditions + ns.partlycloudy|string  %}
                  {% endif %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'cloudy' %}
                  {% set ns.multiplicator = (ns.cloudy/my_forecast['weather.forecast_home'].forecast[hour].cloud_coverage)*100 %}
                  {% if ns.multiplicator <= 1 %}
                  {% set ns.conditions = ns.conditions + (ns.cloudy*ns.multiplicator)|round(2)|string  %}
                  {% else %}
                  {% set ns.conditions = ns.conditions + ns.cloudy|string  %}
                  {% endif %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'fog' %}
                  {% set ns.conditions = ns.conditions + ns.fog|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'hail' %}
                  {% set ns.conditions = ns.conditions + ns.hail|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'rainy' %}
                  {% set ns.conditions = ns.conditions + ns.rainy|string %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'pouring' %}
                  {% set ns.conditions = ns.conditions + ns.pouring|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'windy' %}
                  {% set ns.conditions = ns.conditions + ns.windy|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'windy-variant' %}
                  {% set ns.conditions = ns.conditions + ns.windy|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'snowy' %}
                  {% set ns.conditions = ns.conditions + ns.snowy|string %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'snowy-rainy' %}
                  {% set ns.conditions = ns.conditions + ns.snowy|string  %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'exceptional' %}
                  {% set ns.conditions = ns.conditions + ns.exceptional|string  %}
                {%- else %}
                  {%- set ns.conditions = ns.conditions + 1|string %}
                {%- endif %}
            {%- else %}
                {%- if my_forecast['weather.forecast_home'].forecast[hour].condition == 'partlycloudy' %}
                  {% set ns.multiplicator = (ns.partlycloudy/my_forecast['weather.forecast_home'].forecast[hour].cloud_coverage)*100 %}
                  {% if ns.multiplicator <= 1 %}
                  {% set ns.conditions = ns.conditions + (ns.partlycloudy*ns.multiplicator)|round(2)|string + ns.separator  %}
                  {% else %}
                  {% set ns.conditions = ns.conditions + ns.partlycloudy|string + ns.separator %}
                  {% endif %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'cloudy' %}
                  {% set ns.multiplicator = (ns.cloudy/my_forecast['weather.forecast_home'].forecast[hour].cloud_coverage)*100 %}
                  {% if ns.multiplicator <= 1 %}
                  {% set ns.conditions = ns.conditions + (ns.cloudy*ns.multiplicator)|round(2)|string + ns.separator %}
                  {% else %}
                  {% set ns.conditions = ns.conditions + ns.cloudy|string + ns.separator %}
                  {% endif %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'fog' %}
                  {% set ns.conditions = ns.conditions + ns.fog|string + ns.separator %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'hail' %}
                  {% set ns.conditions = ns.conditions + ns.hail|string + ns.separator %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'rainy' %}
                  {% set ns.conditions = ns.conditions + ns.rainy|string + ns.separator -%}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'pouring' %}
                  {% set ns.conditions = ns.conditions + ns.pouring|string + ns.separator %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'windy' %}
                  {% set ns.conditions = ns.conditions + ns.windy|string + ns.separator -%}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'windy-variant' %}
                  {% set ns.conditions = ns.conditions + ns.windy|string + ns.separator %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'snowy' -%}
                  {% set ns.conditions = ns.conditions + ns.snowy|string + ns.separator -%}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'snowy-rainy' %}
                  {% set ns.conditions = ns.conditions + ns.snowy|string + ns.separator %}
                {%- elif my_forecast['weather.forecast_home'].forecast[hour].condition == 'exceptional' %}
                  {% set ns.conditions = ns.conditions + ns.exceptional|string + ns.separator %}
                {%- else -%}
                  {% set ns.conditions = ns.conditions + 1|string + ns.separator %}
                {%- endif %}
            {%- endif %}
          {%- endfor %}
          {{ ns.conditions }}

- trigger:
    - platform: time_pattern
      hours: /1
  action:
    - service: weather.get_forecasts
      data:
        type: hourly
      target:
        entity_id: weather.forecast_home
      response_variable: hourly
  sensor:
    - name: Weather Forecast Hourly
      unique_id: weather_forecast_hourly
      state: "{{ now().isoformat() }}"
      attributes:
        forecast: "{{ hourly['weather.forecast_home'].forecast }}"
```

This all goes into ``templates.yaml``, or whichever place you have configured to hold your templates. There could be a thousand ways to improve the code, such as:

- run it only for darker parts of the year (depending on hemisphere, that is)
- only calculate the dampening values for hours like 6:00 - 18:00

It runs really well for the darker days. There could be an input boolean to deal with  snow covered panels where duration is >= 1 day, which this sensor fails to deal with.

I have no doubt folks here will take it apart to suit their needs and share the results.

Then I have an automation on top of that to poll the solcast servers at specified times and the result is +/- what I want:

```  
alias: Solcast dampening values update
  description: ''
  trigger:
  - platform: state
    entity_id:
    - sensor.daily_dampening_values_from_hourly_forecast_conditions
    attribute: dampening
  condition: []
  action:
  - service: solcast_solar.set_dampening
    data:
      damp_factor: '{{ state_attr(''sensor.daily_dampening_values_from_hourly_forecast_conditions_2'',
        ''dampening'') }}'
    enabled: true
  - delay:
      seconds: '{{ range(30, 360)|random|int }}'
  - service: solcast_solar.update_forecasts
    metadata: {}
    data: {}
```

I meant to chime in on the discussion about dampening here: https://github.com/BJReplay/ha-solcast-solar/discussions/120 to say the dampening is a good way for some of us to deal with one's location constrains (where a shadow from a chimney is the least of my problems)...

**Reply by zhouska** [reply]
> > It calculates dampening values using `weather.forecast_home` entity for conditions like cloudy, fog, hail, pouring and snowy. I have assigned some manual and arbitrary 'weights' (trial and error) to these conditions to recalculate the dampening values based on the output from the entity.
> 
> This is fantastic.
> 
> If you could, sharing your code could be a great starting point for others to adapt from their forecast as well.

if you find it useful, maybe it could go into this section of the README?

![image](https://github.com/user-attachments/assets/a7aabdad-1ad3-4ee5-855c-f132f0acf709)

as it in fact does exactly what is mentioned there...

**Reply by gcoan** [reply]
> It calculates dampening values using weather.forecast_home entity for conditions like cloudy, fog, hail, pouring and snowy. I have assigned some manual and arbitrary 'weights' (trial and error) to these conditions to recalculate the dampening values based on the output from the entity.

I've been thinking about this and wondering how much the automation is double-counting the work that Solcast already do?

According to Solcast https://solcast.com/forecast-accuracy their forecast takes account of multi-year actual generation figures plus they have near real-time cloud tracking from 11 weather satellites so in theory their solar estimation should be pretty accurate and should already take account of the predicted weather forecast.

From my own experience I think there is merit in doing some calibration of the solcast forecast data against PV actuals, panels and inverters are never going to be 100% efficient.  I found I was consistently getting lower than the Solcast 50% forecast but since I dropped the Solcast efficiency factor from 99% to 85% I've got much better correlation.  I need to do more feedback loop checking but this is giving me good results on a day to day basis
